### ISSUE TITLE
Train on Custom Dataset

### ISSUE THREAD

**Issue by LegendFateX** [SwinTransformer/Swin-Transformer-Semantic-Segmentation#47]
Hi, I really appreciate this wonderful work, but I got some problems on training on my custom dataset. I have a dataset with only 2 classes (0: background; 1:target), and class 0 accounts for 95%+, I just modify the config file as follow:
![2021-10-19_215001](https://user-images.githubusercontent.com/50018479/137923673-438a0a75-9e47-4ed4-9012-1e621266ce17.png)
but mIoU will keep unchanged after a few epochs of training, the info is as follow:
![2021-10-19_215246](https://user-images.githubusercontent.com/50018479/137924495-ed9e337d-8a99-4036-b45f-76fbc37ed9ae.png)
Is there anything that I missed or any configs I overlooked? Hope someone could give me some suggestions, thx in advance!!

**Comment by KeanuWongves** [comment]
> Hi, I really appreciate this wonderful work, but I got some problems on training on my custom dataset. I have a dataset with only 2 classes (0: background; 1:target), and class 0 accounts for 95%+, I just modify the config file as follow: ![2021-10-19_215001](https://user-images.githubusercontent.com/50018479/137923673-438a0a75-9e47-4ed4-9012-1e621266ce17.png) but mIoU will keep unchanged after a few epochs of training, the info is as follow: ![2021-10-19_215246](https://user-images.githubusercontent.com/50018479/137924495-ed9e337d-8a99-4036-b45f-76fbc37ed9ae.png) Is there anything that I missed or any configs I overlooked? Hope someone could give me some suggestions, thx in advance!!

Hi, I was about to train on my own dataset too , and it has 2 classes and the negitive one is above 95% like yours. 
I think we can share some experience for training .
how do you say?

**Comment by LegendFateX** [comment]
> > Hi, I really appreciate this wonderful work, but I got some problems on training on my custom dataset. I have a dataset with only 2 classes (0: background; 1:target), and class 0 accounts for 95%+, I just modify the config file as follow: ![2021-10-19_215001](https://user-images.githubusercontent.com/50018479/137923673-438a0a75-9e47-4ed4-9012-1e621266ce17.png) but mIoU will keep unchanged after a few epochs of training, the info is as follow: ![2021-10-19_215246](https://user-images.githubusercontent.com/50018479/137924495-ed9e337d-8a99-4036-b45f-76fbc37ed9ae.png) Is there anything that I missed or any configs I overlooked? Hope someone could give me some suggestions, thx in advance!!
> 
> Hi, I was about to train on my own dataset too , and it has 2 classes and the negitive one is above 95% like yours. I think we can share some experience for training . how do you say?

Hi, I would like to share with you some of my configs. Except the configs I've mentioned above, I've also tried to change the loss to LovaszLoss, but new bugs erupted, the loss became Nan. And I've tried to increase learning rate to 1e-4, but the train results were just the same as the results I posted above. However, another repository [https://github.com/JIA-HONG-CHU/Swin-Transformer-add-EncNet-DaNet-DraNet-for-semantic-segmentation-on-Statelite-Dataset](url) showed that Swin Transformer can be finetuned on the datasets which are similar to what I've used. How about your training process? : )

**Comment by KeanuWongves** [comment]
> > > Hi, I really appreciate this wonderful work, but I got some problems on training on my custom dataset. I have a dataset with only 2 classes (0: background; 1:target), and class 0 accounts for 95%+, I just modify the config file as follow: ![2021-10-19_215001](https://user-images.githubusercontent.com/50018479/137923673-438a0a75-9e47-4ed4-9012-1e621266ce17.png) but mIoU will keep unchanged after a few epochs of training, the info is as follow: ![2021-10-19_215246](https://user-images.githubusercontent.com/50018479/137924495-ed9e337d-8a99-4036-b45f-76fbc37ed9ae.png) Is there anything that I missed or any configs I overlooked? Hope someone could give me some suggestions, thx in advance!!
> > 
> > 
> > Hi, I was about to train on my own dataset too , and it has 2 classes and the negitive one is above 95% like yours. I think we can share some experience for training . how do you say?
> 
> Hi, I would like to share with you some of my configs. Except the configs I've mentioned above, I've also tried to change the loss to LovaszLoss, but new bugs erupted, the loss became Nan. And I've tried to increase learning rate to 1e-4, but the train results were just the same as the results I posted above. However, another repository [https://github.com/JIA-HONG-CHU/Swin-Transformer-add-EncNet-DaNet-DraNet-for-semantic-segmentation-on-Statelite-Dataset](url) showed that Swin Transformer can be finetuned on the datasets which are similar to what I've used. How about your training process? : )

I have just look into the codes, cuz I'm not familiar with mmcv so I have no imformation for u right now, and my photos resolution are not like pre-trained 512X512, and it's like infrared, I think I will trian from scratch. Thanks for your kindly share , I'll share my exp with you if I get any progress.

**Comment by VV1314** [comment]
For your dataset with only two classes, maybe you can change "use_sigmoid=False" into "use_sigmoid=True". However, when I train with my own dataset also with two classes, loss always keeps zero

**Comment by LegendFateX** [comment]
> For your dataset with only two classes, maybe you can change "use_sigmoid=False" into "use_sigmoid=True". However, when I train with my own dataset also with two classes, loss always keeps zero

I've also tried your configs, and I've confronted 0-loss problem, too.

**Comment by VV1314** [comment]
So you have handled it?Could you tell me how to solve the problem?Thank you!

**Comment by TNA8** [comment]
I also have custom dataset with 2 classes.
Could you help me?

* What config file did you use?
* What is the format and directory structure of dataset for `ADE20KDataset`?
I have pairs of input image and masked image like this.

![screenshot_9](https://user-images.githubusercontent.com/58914448/140319637-88ef48b2-c903-4ba4-8303-7bb917cb96ff.png)

Thanks in advance.

**Comment by KeanuWongves** [comment]
> > For your dataset with only two classes, maybe you can change "use_sigmoid=False" into "use_sigmoid=True". However, when I train with my own dataset also with two classes, loss always keeps zero
> 
> I've also tried your configs, and I've confronted 0-loss problem, too.

I tried use focal_loss on my own dataset and it get some improved

**Comment by Jay-IPL** [comment]
If you have 2 classes background and target, and you use an ade-type custom dataset, I think num_classes = 1 and sigmoid=True.

**Comment by manaclan** [comment]
I got the same problem with my dataset (ade20k config), after changing to `reduce_zero_label=False` everything worked.

**Comment by Jay-IPL** [comment]
> I got the same problem with my dataset (ade20k config), after changing to `reduce_zero_label=False` everything worked.

Did you also set num_classes = 1 and sigmoid=True?

**Comment by manaclan** [comment]
@Jay-IPL Only num_classes = 2, sigmoid left untouch. Here is my config:
```
model = dict(
    backbone=dict(
        embed_dim=128,
        depths=[2, 2, 18, 2],
        num_heads=[4, 8, 16, 32],
        window_size=7,
        ape=False,
        drop_path_rate=0.3,
        patch_norm=True,
        use_checkpoint=False
    ),
    decode_head=dict(
        in_channels=[128, 256, 512, 1024],
        num_classes=2,
        loss_decode=dict(
            type='CrossEntropyLoss', 
            use_sigmoid=False, 
            loss_weight=1.0,
            class_weight=[0.01, 1])
    ),
    auxiliary_head=dict(
        in_channels=512,
        num_classes=2
    ))
```

**Comment by Jay-IPL** [comment]
> 

got it. But in custom_dataset.py did you set CLASSES = ('background','target') or CLASSES = (target')?

**Comment by Jay-IPL** [comment]
> @Jay-IPL Only num_classes = 2, sigmoid left untouch. Here is my config:
> 
> ```
> model = dict(
>     backbone=dict(
>         embed_dim=128,
>         depths=[2, 2, 18, 2],
>         num_heads=[4, 8, 16, 32],
>         window_size=7,
>         ape=False,
>         drop_path_rate=0.3,
>         patch_norm=True,
>         use_checkpoint=False
>     ),
>     decode_head=dict(
>         in_channels=[128, 256, 512, 1024],
>         num_classes=2,
>         loss_decode=dict(
>             type='CrossEntropyLoss', 
>             use_sigmoid=False, 
>             loss_weight=1.0,
>             class_weight=[0.01, 1])
>     ),
>     auxiliary_head=dict(
>         in_channels=512,
>         num_classes=2
>     ))
> ```

got it. But in custom_dataset.py did you set CLASSES = ('background','target') or CLASSES = (target')?

**Comment by KeanuWongves** [comment]
> > > 
> > 
> > 
> > got it. But in custom_dataset.py did you set CLASSES = ('background','target') or CLASSES = (target')?
> 
> > @Jay-IPL Only num_classes = 2, sigmoid left untouch. Here is my config:
> > ```
> > model = dict(
> >     backbone=dict(
> >         embed_dim=128,
> >         depths=[2, 2, 18, 2],
> >         num_heads=[4, 8, 16, 32],
> >         window_size=7,
> >         ape=False,
> >         drop_path_rate=0.3,
> >         patch_norm=True,
> >         use_checkpoint=False
> >     ),
> >     decode_head=dict(
> >         in_channels=[128, 256, 512, 1024],
> >         num_classes=2,
> >         loss_decode=dict(
> >             type='CrossEntropyLoss', 
> >             use_sigmoid=False, 
> >             loss_weight=1.0,
> >             class_weight=[0.01, 1])
> >     ),
> >     auxiliary_head=dict(
> >         in_channels=512,
> >         num_classes=2
> >     ))
> > ```
> 
> got it. But in custom_dataset.py did you set CLASSES = ('background','target') or CLASSES = (target')?

But, I think use num_classes = 1 with sigmoid is the same when you using num_classes= 2 and with softmax , It's basiclly the same in mathematics

**Comment by Jay-IPL** [comment]
> > > > 
> > > 
> > > 
> > > got it. But in custom_dataset.py did you set CLASSES = ('background','target') or CLASSES = (target')?
> > 
> > 
> > > @Jay-IPL Only num_classes = 2, sigmoid left untouch. Here is my config:
> > > ```
> > > model = dict(
> > >     backbone=dict(
> > >         embed_dim=128,
> > >         depths=[2, 2, 18, 2],
> > >         num_heads=[4, 8, 16, 32],
> > >         window_size=7,
> > >         ape=False,
> > >         drop_path_rate=0.3,
> > >         patch_norm=True,
> > >         use_checkpoint=False
> > >     ),
> > >     decode_head=dict(
> > >         in_channels=[128, 256, 512, 1024],
> > >         num_classes=2,
> > >         loss_decode=dict(
> > >             type='CrossEntropyLoss', 
> > >             use_sigmoid=False, 
> > >             loss_weight=1.0,
> > >             class_weight=[0.01, 1])
> > >     ),
> > >     auxiliary_head=dict(
> > >         in_channels=512,
> > >         num_classes=2
> > >     ))
> > > ```
> > 
> > 
> > got it. But in custom_dataset.py did you set CLASSES = ('background','target') or CLASSES = (target')?
> 
> But, I think use num_classes = 1 with sigmoid is the same when you using num_classes= 2 and with softmax , It's basiclly the same in mathematics

Yes, you are right. But in original mmseg/datasets/ade.py, 'background' is not in CLASSES, CLASSES have 150 targets, and in ade config file num_classes=150. So, my question is that if you use  num_classes=2 in your config file, your CLASSES should only have 1 target, right?

**Comment by KeanuWongves** [comment]
> > > > > 
> > > > 
> > > > 
> > > > got it. But in custom_dataset.py did you set CLASSES = ('background','target') or CLASSES = (target')?
> > > 
> > > 
> > > > @Jay-IPL Only num_classes = 2, sigmoid left untouch. Here is my config:
> > > > ```
> > > > model = dict(
> > > >     backbone=dict(
> > > >         embed_dim=128,
> > > >         depths=[2, 2, 18, 2],
> > > >         num_heads=[4, 8, 16, 32],
> > > >         window_size=7,
> > > >         ape=False,
> > > >         drop_path_rate=0.3,
> > > >         patch_norm=True,
> > > >         use_checkpoint=False
> > > >     ),
> > > >     decode_head=dict(
> > > >         in_channels=[128, 256, 512, 1024],
> > > >         num_classes=2,
> > > >         loss_decode=dict(
> > > >             type='CrossEntropyLoss', 
> > > >             use_sigmoid=False, 
> > > >             loss_weight=1.0,
> > > >             class_weight=[0.01, 1])
> > > >     ),
> > > >     auxiliary_head=dict(
> > > >         in_channels=512,
> > > >         num_classes=2
> > > >     ))
> > > > ```
> > > 
> > > 
> > > got it. But in custom_dataset.py did you set CLASSES = ('background','target') or CLASSES = (target')?
> > 
> > 
> > But, I think use num_classes = 1 with sigmoid is the same when you using num_classes= 2 and with softmax , It's basiclly the same in mathematics
> 
> Yes, you are right. But in original mmseg/datasets/ade.py, 'background' is not in CLASSES, CLASSES have 150 targets, and in ade config file num_classes=150. So, my question is that if you use num_classes=2 in your config file, your CLASSES should only have 1 target, right?

probably u r right, it make sense.

**Comment by manaclan** [comment]
> id left untou

I used `classes=['background','target']`

**Comment by lixixin** [comment]
> 
![image](https://user-images.githubusercontent.com/85006440/146720760-14ec5562-3f5f-4308-b0f1-9051b1e080d3.png)

**Comment by SRN200722** [comment]
> For your dataset with only two classes, maybe you can change "use_sigmoid=False" into "use_sigmoid=True". However, when I train with my own dataset also with two classes, loss always keeps zero

Have you solved the problem about loss always keeping zero? I meet the same problem when I train my own dataset.

**Comment by ojipadeson** [comment]
> > For your dataset with only two classes, maybe you can change "use_sigmoid=False" into "use_sigmoid=True". However, when I train with my own dataset also with two classes, loss always keeps zero
> 
> Have you solved the problem about loss always keeping zero? I meet the same problem when I train my own dataset.

Have you find any method to improve the results? I meet the same problem.
